bdk_chain/src/tx_graph.rs

```diff
@@ -1,58 +1,69 @@
 use alloc::vec::Vec;
 use bitcoin::{OutPoint, Transaction, TxOut, Txid};
+use core::ops::RangeInclusive;
 
-#[derive(Clone, Debug, Default, PartialEq)]
-pub struct TxGraph {
-    txs: HashMap<Txid, TxNode>,
+#[derive(Clone, Debug, PartialEq)]
+pub struct TxGraph<T = Transaction> {
+    txs: HashMap<Txid, TxNode<T>>,
     spends: BTreeMap<OutPoint, HashSet<Txid>>,
 
     // This atrocity exists so that `TxGraph::outspends()` can return a reference.
     // FIXME: This can be removed once `HashSet::new` is a const fn.
     empty_outspends: HashSet<Txid>,
 }
 
+impl<T> Default for TxGraph<T> {
+    fn default() -> Self {
+        Self {
+            txs: Default::default(),
+            spends: Default::default(),
+            empty_outspends: Default::default(),
+        }
+    }
+}
+
 /// Node of a [`TxGraph`]
 #[derive(Clone, Debug, PartialEq)]
-enum TxNode {
-    Whole(Transaction),
+enum TxNode<T = Transaction> {
+    Whole(T),
     Partial(BTreeMap<u32, TxOut>),
 }
 
-impl Default for TxNode {
+impl<T> Default for TxNode<T> {
     fn default() -> Self {
         Self::Partial(BTreeMap::new())
     }
 }
 
-impl TxGraph {
-    /// The transactions spending from this output.
-    ///
-    /// `TxGraph` allows conflicting transactions within the graph. Obviously the transactions in
-    /// the returned will never be in the same blockchain.
-    pub fn outspends(&self, outpoint: OutPoint) -> &HashSet<Txid> {
-        self.spends.get(&outpoint).unwrap_or(&self.empty_outspends)
+impl<T: AsTransaction> TxGraph<T> {
+    /// Iterate over all tx outputs known by [`TxGraph`].
+    pub fn all_txouts(&self) -> impl Iterator<Item = (OutPoint, &TxOut)> {
+        self.txs.iter().flat_map(|(txid, tx)| match tx {
+            TxNode::Whole(tx) => tx
+                .as_tx()
+                .output
+                .iter()
+                .enumerate()
+                .map(|(vout, txout)| (OutPoint::new(*txid, vout as _), txout))
+                .collect::<Vec<_>>(),
+            TxNode::Partial(txouts) => txouts
+                .iter()
+                .map(|(vout, txout)| (OutPoint::new(*txid, *vout as _), txout))
+                .collect::<Vec<_>>(),
+        })
     }
 
-    /// The transactions spending from `txid`.
-    pub fn tx_outspends(
-        &self,
-        txid: Txid,
-    ) -> impl DoubleEndedIterator<Item = (u32, &HashSet<Txid>)> + '_ {
-        let start = OutPoint { txid, vout: 0 };
-        let end = OutPoint {
-            txid,
-            vout: u32::MAX,
-        };
-        self.spends
-            .range(start..=end)
-            .map(|(outpoint, spends)| (outpoint.vout, spends))
+    /// Iterate over all full transactions in the graph.
+    pub fn full_transactions(&self) -> impl Iterator<Item = &Transaction> {
+        self.txs.iter().filter_map(|(_, tx)| match tx {
+            TxNode::Whole(tx) => Some(tx.as_tx()),
+            TxNode::Partial(_) => None,
+        })
```

Comment:
Nit: I think it will make the API more consistent if this is an iterator of `&T`.

```suggestion
    pub fn full_transactions(&self) -> impl Iterator<Item = &T> {
        self.txs.iter().filter_map(|(_, tx)| match tx {
            TxNode::Whole(tx) => Some(tx),
            TxNode::Partial(_) => None,
        })
```